# Homework 4

This notebook provides some skeleton code to get you started on the homework.  Add in your own code and markdown cells  to answer the homework questions.  You will submit this notebook as an .ipynb file.  

Before starting, select "Runtime->Factory reset runtime" to start with your directories and environment in the base state.

If you want to save changes to the notebook, select "File->Save a copy in Drive" from the top menu in Colab.  This will save the notebook in your Google Drive.

# Clones, Installs, and Imports

## Clone Repo

In [ ]:
!git clone https://github.com/zlisto/social_media_genAI

import os
os.chdir("social_media_genAI/main")


Cloning into 'social_media_genAI'...
remote: Enumerating objects: 2059, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 2059 (delta 4), reused 8 (delta 4), pack-reused 2049 (from 4)
Receiving objects: 100% (2059/2059), 279.17 MiB | 29.66 MiB/s, done.
Resolving deltas: 100% (188/188), done.
Updating files: 100% (1624/1624), done.


## Installs

In [ ]:
!pip install -q -r requirements.txt

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 55.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 548.3/548.3 kB 34.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 63.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.0/85.0 kB 6.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 232.6/232.6 kB 17.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 244.3/244.3 kB 18.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 294.7/294.7 kB 21.7 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
nx-cugraph-cu12 24.12.0 requires networkx>=3.2, but you have networkx 3.0 which is incompatible.
torch 2.5.1+cu124 requires nvidia-cublas-cu12==12.4.5.8; platform_system == "Linux" and platform_machine == "x86_64", but you have 

## Imports

In [ ]:
# Standard library imports
import os
import time
import math
import random
import ast
import base64
import subprocess
import datetime
from datetime import datetime, timedelta
from io import BytesIO
import textwrap

# Third-party imports
import pandas as pd
import numpy as np
import requests
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import cv2  # OpenCV for video processing
from tqdm import tqdm
from PIL import Image as PILImage
from IPython.display import display, Image, HTML, Audio, Video
from moviepy.editor import ImageClip, AudioFileClip, concatenate_videoclips
import openai

# Google Colab-specific imports
from google.colab import userdata

#helper functions for using ChatGPT
from scripts.genai import GenAI

#this option makes it so tweets display nicely in a dataframe
pd.set_option("display.max_colwidth", None)

#this code sets the font sizes for plots
plt.rcParams.update({'axes.labelsize': 18,
                     'xtick.labelsize': 14,
                     'ytick.labelsize': 14,
                     'figure.figsize':(8,6),
                     'axes.grid':True})


  if event.key is 'enter':



## Connect to ChatGPT

In [ ]:
#set your OpenAI API key as a string
OPENAI_API_KEY = ""


In [ ]:
# Initialize GenAI
jarvis = GenAI(OPENAI_API_KEY)

text = "Who are you?"
response = jarvis.generate_text(text)
print(f"Human:{text}\nAI:{response}")

Human:Who are you?
AI:I am Jarvis, an AI designed to assist you with a variety of tasks, answer questions, and provide information based on a vast amount of knowledge I have been trained on. How can I help you today?


# Problem 1 (20 points) Basic Image Analysis

## Problem 1.1 (5 points) View image

Load the image with file name `"data/image_compressed_taylorswift/1838671987862472764_11830955_small.jpeg"` and display it with the `Image` function.

## Problem 1.2 (10 points) Instagram caption recommendations

Use the AI to provide 3 captions for this image on Instagram to make the image more engaging for people who are fans of Metallica and explain why the captions are engaging.  Tell the AI to return the result as HTML with big font and modern layout.  Display the captions with the `HTML` function.

## Problem 1.3 (5 points) Rolling Stone article

Put the following images into a list called `image_paths`:
1. `"/content/social_media_genAI/main/data/image_compressed_taylorswift/1661245127015635399_11830955_small.jpeg"`

2.`"/content/social_media_genAI/main/data/image_compressed_taylorswift/3144002220890559101_11830955_small.jpeg"`

3. `"/content/social_media_genAI/main/data/image_compressed_taylorswift/3235298520231318519_11830955_small.jpeg"`

Display the images with `display` and  `Image` functions.  

Then have the AI analyze these images simulataneously and describe their style evolution.  Return the answer as an article for Rolling Stone with a clever title about Taylor Swift, in nice html.  Display the article with the `HTML` function.


# Problem 2 (20 points) Tweeting a Reaction to an Image

In this problem you will have the AI tweet a reaction to an image in the style of a specific person.


## Problem 2.1 (5 points) Load Image

Load the image with file name `"data/image_compressed_cristiano/1603695572069767264_173560420_small.jpeg"` and display it with the `Image` function.


## Problem 2.2 (7 points) Describe Twitter user

Write a `user_description` of a Twitter user who will see this image and react to it.  Be creative :).  Then give the AI `instructions` that tell it to provide more details about the tweeting style of this user as nice HTML.  Call this output string `tweet_style_html`.  Display the tweet style using `HTML`.

## Problem 2.3 (8 points) Tweet reaction to image

Use the AI to write a tweet in the style of the user from the previous part (`tweet_style_html`) to the image from part 1.  Display the image and then display your tweet using the `display_tweet` function.


# Problem 3. (30 points) Generate Text Narration of a Video

## Problem 3.1 (5 points) Load video

Load the video located at `video_path = "data/videos/superbon_banchamek - boxing.mp4"` and display it with the `Video` function.  Make sure the video is visible.

## Problem 3.2 (5 points) Narration length

A human speaks 200 words per minute.  Using this, calculate `nwords_max`, the maximum number of words in the narration for the video.   Display the duration of the video in seconds and display `nwords_max`.

## Problem 3.3 (10 points) Create text narration for video

Write a `narration` for the video in the voice of the fighter, whose name is Superbon (he is the fighter in the video and this is his gym).  Make your narration target Western UFC fans to take a beginners class in his gym.  The gym's name is Superbon Training Camp.

  First have the AI give you a short `bio`of Superbon and his gym, formatted as a nice HTML.  Display the `bio` with the `display` and `HTML` functions.  

Then give this `bio` as part of you `instructions` for the video `narration`.  Make sure the AI only returns the narration and nothing else.

 Make sure the `narration` has fewer than `nwords_max` words.  Display the `narration`.

## Problem 3.4 (4 points) Word count of narration

Display the number of words in the narration along with `nwords_max`.  Make sure the number of words is less than `nwords_max`. If it does not, generate another narration.

## Problem 3.5 (6 points) Evaluate narration

Have the AI evaluate on a scale of 1 to 10, how effective this narration and this video will be in attracting the Western UFC fan to visit the gym and become a paying customer.  Return the evaluation as an HTML that is easy to read.  Use the `generate_video_description` function, and include the `bio` and `narration` in the `instructions`.  Display the evalutaion with the `HTML` and `display` functions.

# Problem 4 (30 points) Image Database

We will create an indexed image data base and then try to build an Instagram reel from it.

## Problem 4.1 (15 points) Create index

The images are in the folder `"data/officialcaptainamerica"`.  Have the AI generate a description of 10 images in the folder.  Only do 10 images and no more as it will take 10 minutes to do all the images in the folder.   

Make sure your `instructions` to the AI tell it this is an image from the Instagram account of the new movie Captain America: Brave New World starring Anthony Mackey as the new Captain America.

Create a dataframe `df` with columns `"image_path"` and `"image_description"` that has the full image path and an AI generated description for each of the 10 images.  

Display the head of `df`.




## Problem 4.2 (10 points) Image slide show

Politely ask the AI to generate a 3 scene reel to promote Captain America: Brave New World, with the image path and a narration for each scene, selecting from the images in `df`. The narration should be done by Steve Rogers, the original Captain America, who is talking about the new Captain America played by Anthony Mackey.

Have the AI return a JSON object with structure:

`{'scene_list':[{'image_path': 'path/to/image', 'narration': 'narration for the scene'}, ...]}`.  

Make sure the AI `output_type = "json_object"`.  

Then use the `ast.literal_eval` and `pd.DataFrame` functions to convert the `'scene_list'` in the JSON object into a dataframe called `df_script`.  

Dispay the head of `df_script`.

## Problem 4.3 (5 points) Display images and narrations

Plot the images in `df_script` in a single horizontal row and put the narration below the images. Make sure the text for each scene is contained under its image.  The `textwrap.wrap` function is useful here.